File: Classes/GBFlatButton.m
//
//  GBFlatButton.m
//  GBFlatButtonExample
//
//  Created by Gustavo Barbosa on 4/15/14.
//  Copyright (c) 2014 Gustavo Barbosa. All rights reserved.
//

#import "GBFlatButton.h"
#import <QuartzCore/QuartzCore.h>

static CGFloat const kHorizontalPadding = 14.0f;

@implementation GBFlatButton

#pragma mark - Constructors

- (id)init
{
    self = [super init];
    if (self) {
        [self customizeAppearance];
    }
    return self;
}

- (id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self customizeAppearance];
    }
    return self;
}

- (id)initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    if (self) {
        [self customizeAppearance];
    }
    return self;
}

#pragma mark - Appearance

- (void)customizeAppearance
{
    BOOL containsEdgeInsets = ! UIEdgeInsetsEqualToEdgeInsets(self.contentEdgeInsets, UIEdgeInsetsZero);
    self.contentEdgeInsets = containsEdgeInsets ? self.contentEdgeInsets : UIEdgeInsetsMake(0, kHorizontalPadding, 0, kHorizontalPadding);
    self.layer.borderWidth = self.layer.borderWidth ?: 1.0f;
    self.layer.cornerRadius = self.layer.cornerRadius ?: CGRectGetHeight(self.frame) / 2.0f;
    self.clipsToBounds = YES;
}

- (void)setSelected:(BOOL)selected
{
    [super setSelected:selected];
    
    if (selected)
        self.backgroundColor = self.buttonColor;
    else
        self.backgroundColor = [UIColor whiteColor];
}

- (void)drawRect:(CGRect)rect
{
    self.layer.borderColor = self.buttonColor.CGColor;
    [self setTitleColor:self.buttonColor
               forState:UIControlStateNormal];
    [self setTitleColor:[UIColor whiteColor]
               forState:UIControlStateSelected];
}

#pragma mark - Acessors

- (UIColor *)buttonColor
{
    return _buttonColor ?: self.tintColor;
}

@end

File: Classes/swift/GBFlatButton.swift
//
//  GBFlatButton.swift
//  GBFlatButton
//
//  Created by Gustavo Barbosa on 6/6/14.
//  Copyright (c) 2014 Gustavo Barbosa. All rights reserved.
//

import UIKit

class GBFlatButton : UIButton {
    
    let kHorizontalPadding : Float = 14.0
    
    var buttonColor: UIColor?
    
    init(frame: CGRect) {
        super.init(frame: frame)
        _customizeAppearance()
    }
    
    init(coder aDecoder: NSCoder!) {
        super.init(coder: aDecoder)
        _customizeAppearance()
    }
    
    override func drawRect(rect: CGRect) {
        self.layer.borderColor = self.tintColor.CGColor
        self.setTitleColor(self.tintColor, forState: UIControlState.Normal)
        self.setTitleColor(UIColor.whiteColor(), forState: UIControlState.Selected)
    }
    
    func _customizeAppearance() {
        
        let containsEdgeInsets = !UIEdgeInsetsEqualToEdgeInsets(contentEdgeInsets, UIEdgeInsetsZero)
        contentEdgeInsets = containsEdgeInsets ? contentEdgeInsets : UIEdgeInsets(top: 0, left: kHorizontalPadding, bottom: 0, right: kHorizontalPadding)
        
        if layer.borderWidth == 0.0 {
            layer.borderWidth = 1.0
        }
        
        layer.borderColor = tintColor.CGColor
        
        if layer.cornerRadius == 0.0 {
            layer.cornerRadius = frame.size.height / 2.0
        }

//        layer.masksToBounds = true
    }
    
    override var selected: Bool {
        get {
            return super.selected
        }
        set {
            super.selected = newValue
            if super.selected {
                self.backgroundColor = self.tintColor
            } else {
                self.backgroundColor = UIColor.whiteColor()
            }
        }
    }
    
    override var tintColor: UIColor! {
        get {
            if let bc = self.buttonColor {
                return bc
            } else {
                return super.tintColor
            }
        }
        set {
            super.tintColor = newValue
            self.buttonColor = newValue
        }
    }
}